lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -856,199 +872,242 @@ void mlir::triton::updateWaits(ModuleOp module) {
+  if (auto dotOp = dyn_cast<tt::DotOp>(op)) {
+    auto resEnc =
+        dotOp.getType().getEncoding().dyn_cast<ttg::NvidiaMmaEncodingAttr>();
+    if (resEnc && resEnc.isHopper()) {
+      return dotOp;
     }
   }
-  return false;
+  return nullptr;
 }
 
-static void removeExtraWait(tt::nvidia_gpu::DotWaitOp dotWaitOp,
-                            bool hasDotWait0) {
-  if (hasDotWait0) {
-    dotWaitOp->erase();
+// Determines whether a tt.dot op can be transformed into ttng.dot.async.  The
+// dot can be made async if all of the following are true.
+//
+//  0. The operation is an MMAv3 dot.
+//
+//  1. All operands that touch shared memory are multi-buffered, i.e. can't read
+//     an incomplete value while it's being written asynchronously by a load.
+//
+//  2. During iteration i, nothing other than the loop's `yield` reads the
+//     result of the dot.
+//
+//  3. During iteration i, the result of the dot from iteration i-1 is consumed
+//     only by other MMAv3 dots (either sync or async) as the `c` operand.
+//
+//     This is safe because the following pseudo-PTX is valid (even though the
+//     docs don't say so explicitly):
+//
+//        %accum = dot.async %a1, %b1, %c1
+//        %accum = dot.async %a2, %b2, %accum
+//
+//     That is, the second async dot can use the result of the first one without
+//     an intervening wait.  However, the only operation that can legally read
+//     %accum before the wait is another dot.async, and our current
+//     understanding is this only works for the `c` operand, not `a` or `b`.
+//
+//     A synchronous MMAv3 dot will be codegen'ed as `dot.async; wait 0`, so it
+//     works just as well as an async dot for these purposes.
+//
+// On success, this function returns the index of the dot in the loop's
+// iter_args.  (Rule (2) above ensures this is well-defined.)
+static std::optional<int> dotCanBeMadeAsync(Operation *op, scf::ForOp forOp) {
+  // Rule 0: The operation is an MMAv3 dot.
+  tt::DotOp dotOp = dynCastMMAv3Dot(op);
+  if (!dotOp)
+    return std::nullopt;
+
+  // Rule 1: All shmem operands are used "safely".
+  auto checkOperand = [&](Value operand) {
+    // If it's a shmem operand, it must either be defined outside the loop, or
+    // come from an MemDescSubview op.  Only ConvertLayout and Trans ops are
+    // allowed in between.
+    if (isa<ttg::SharedEncodingAttr>(
+            operand.getType().cast<TensorOrMemDesc>().getEncoding())) {
+      Value transitiveOperand = operand;
+      while (isa_and_nonnull<ttg::ConvertLayoutOp, tt::TransOp>(
+          transitiveOperand.getDefiningOp())) {
+        transitiveOperand = transitiveOperand.getDefiningOp()->getOperand(0);
+      }
+      return forOp.isDefinedOutsideOfLoop(transitiveOperand) ||
+             isa<ttg::MemDescSubviewOp>(transitiveOperand.getDefiningOp());
+    }
+
+    return true;
+  };
+
+  if (!checkOperand(dotOp.getA()) || !checkOperand(dotOp.getB()) ||
+      !checkOperand(dotOp.getC()))
```

Comment:
technically we only care about A and B operand. Are you adding C for completeness? I guess it is always a MMA reg

Reply:
Good point, fixed.